include/deal.II/base/vectorization.h

```diff
@@ -3941,6 +3941,167 @@ operator<<(std::ostream &out, const VectorizedArray<Number, width> &p)
   return out;
 }
 
+
+/**
+ * @name Ternary operations on VectorizedArray
+ */
+//@{
+
+/**
+ * TODO
+ */
+enum SIMDComparison : int
+{
+  equal                 = _CMP_EQ_OQ,
+  not_equal             = _CMP_NEQ_OQ,
+  less_than             = _CMP_LT_OQ,
+  less_than_or_equal    = _CMP_LE_OQ,
+  greater_than          = _CMP_GT_OQ,
+  greater_than_or_equal = _CMP_GE_OQ
+};
+
+
+/**
+ * TODO
+ */
+template <SIMDComparison predicate, typename Number>
```

Comment:
why use a template parameter instead of a regular run-time argument?

Reply:
Not without writing a lot of case statements. (Even though it doesn't look like from the signature,) the mask operations require a compile-time constant expression because the mask operations are reduced to different machine instructions (depending on the int).

Reply:
I don't think I quite understand why the switch/case statements are necessary. You pass the `predicate` value to `_mm512_cmp_ps`, which as far as I can see takes an `int` as last argument. Why does it have to be a compile-time constant?

Reply:
The underlying problem is that the `_mm256_cmp_ps` intrinsic is a bit of a fake function that reduces the last argument to a dedicated machine instruction.

So, considering the following:
```
#include <x86intrin.h>
inline __m256 foo(const int pred, const __m256 left, const __m256 right) {
  return _mm256_cmp_ps(left, right, pred);
}
```

Clang does not like it at all:
```
test.cc:4:10: error: argument to '__builtin_ia32_cmpps256' must be a constant integer
  return _mm256_cmp_ps(left, right, pred);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/llvm/8/bin/../../../../lib/clang/8.0.1/include/avxintrin.h:1840:11: note: expanded from macro '_mm256_cmp_ps'
  (__m256)__builtin_ia32_cmpps256((__v8sf)(__m256)(a), \
          ^
1 error generated.
```

And gcc (after triggering the actual use):
```
test.cc: In function ‘__m256 foo(int, __m256, __m256)’:
test.cc:4:10: error: the last argument must be a 5-bit immediate
    4 |   return _mm256_cmp_ps(left, right, pred);
      |          ^~~~~~~~~~~~~
```

Reply:
I can confirm @tamiko's statement here - there are a considerable number of intrinics which take an `int` argument when it actually needs to be an 8 bit immediate, i.e., compile time constant.

Reply:
I see. How unfortunate. If you don't mind, explain this in two sentences in the (still to be written) documentation so that nobody gets the idea that this could be simplified in the future.